### PR TITLE
Extend Restart to Adaptive samplers

### LINKED ISSUE #1280: [TASK] Improvement on Restart Capability
--------
Issue Description
--------
There are currently several Samplers that do not have the restart capabilities. One reason for that is we do not call the Sampler.finalizeActualSampling(jobHandler,model,myInput) during the restart. The main issue for this is that we need to pass jobHandler to Sampler.finalizeActualSampling, which is not available during restart process. One suggestion is to change it to dict that we can create during restart.

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #1280

##### What are the significant changes in functionality due to this change request?
Extends expected existing restart mechanics to Adaptive samplers. This can be improved once:
- the Sampler's `generateInput` returns a realization instead of storing on `self`
- we have the Realization instance

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by PaulTalbot-INL** [comment]
@wangcj05 @alfoa for your reviewing pleasure.

Note that the Adaptive MonteCarlo did work fine with the restart, as we believed. You have to look at the solution export, especially `exp_restart`, which before this fix only had entries 6-15, and now has 1-15.

**Comment by PaulTalbot-INL** [comment]
It shows me as deleting HERON; is heron a submodule? It shouldn't be until #1271 is merged ...

**Comment by alfoa** [comment]
Test invalidated to get the new modifications in the BasicStats tests (see PR #1282)

File: tests/framework/Samplers/Restart/adaptive_mc.xml [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+<Simulation verbosity="debug">
+  <TestInfo>
+    <name>framework/Samplers/Restart.AdaptiveMC</name>
+    <author>talbpaul</author>
+    <created>2020-07-16</created>
+    <classesTested>Samplers.AdaptiveMonteCarlo</classesTested>
+    <description>
+      Tests restarting an Adaptive Monte Carlo sampling from restart.  \texttt{makeCoarse} samples initial data, then \texttt{makeRestart}
+      makes additional samples, restarting from the first set of samples.  \texttt{makeFine} does all the samples without restart
+      for comparison.  The model for "coarse" always returns a value of 1, while the model for "restart" returns a value of 2, so
```
Comment (wangcj05) [inline tests/framework/Samplers/Restart/adaptive_mc.xml:11]: use ``'' instead ""?

File: tests/framework/Samplers/Restart/adaptive_mc.xml [hunk 2]
```diff
@@ -0,0 +1,196 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="debug">
+  <TestInfo>
+    <name>framework/Samplers/Restart.AdaptiveMC</name>
```
Comment (wangcj05) [inline tests/framework/Samplers/Restart/adaptive_mc.xml:4]: change AdaptiveMC to AMC

File: framework/Runners/PassthroughRunner.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Created on July 17, 2020
+"""
+#for future compatibility with Python 3--------------------------------------------------------------
+from __future__ import division, print_function, unicode_literals, absolute_import
+#End compatibility block for Python 3----------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply framework/Runners/PassthroughRunner.py:19]: I believe the linter will error if it doesn't find this. We should change this globally, as a separate action.

File: framework/Runners/PassthroughRunner.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    A runner for when we already have the answer, but need to go through the mechanics.
+  """
+  def __init__(self, messageHandler, data, metadata=None, uniqueHandler="any", profile=False):
+    """
+      Init method
+      @ In, messageHandler, MessageHandler object, the global RAVEN message
+        handler object
+      @ In, args, dict, this is a list of arguments that will be passed as
+        function parameters into whatever method is stored in functionToRun.
+        e.g., functionToRun(*args)
```
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply framework/Runners/PassthroughRunner.py]: done

File: tests/framework/Samplers/Restart/adaptive_mc.xml (hunk 1 above)
Comment (PaulTalbot-INL) [reply tests/framework/Samplers/Restart/adaptive_mc.xml]: done, used texttt instead

File: tests/framework/Samplers/Restart/adaptive_mc.xml (hunk 2 above)
Comment (PaulTalbot-INL) [reply tests/framework/Samplers/Restart/adaptive_mc.xml]: done

File: framework/Runners/PassthroughRunner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wangcj05) [reply framework/Runners/PassthroughRunner.py:19]: @alfoa I think you are trying to remove the ```__future__``` import, right?

**Comment by wangcj05** [comment]
@alfoa Changes are good to me. Do you want to have review on this PR?

File: framework/Runners/PassthroughRunner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alfoa) [reply framework/Runners/PassthroughRunner.py:19]: Yes we removed it. we can remove the "future" (I know we need to do that to the other files too...I will take care of it)

**Comment by alfoa** [comment]
Comments addressed...

Checklist passed...

PR can be merged

**Pull request merged**